File: oolab/src/main/java/agh/ics/oop/model/Animal.java
package agh.ics.oop.model;

import java.util.Objects;

public class Animal implements WorldElement {
    private MapDirection orientation;
    private Vector2d position;

    public MapDirection getOrientation() {
        return orientation;
    }

    protected void setOrientation(MapDirection orientation) {
        this.orientation = orientation;
    }

    @Override
    public Vector2d getPosition(){
        return position;
    }

    private void setPosition(Vector2d newPosition){
        if(newPosition != null)
            position = newPosition;
    }

    public Animal(Vector2d position){
        this.position = position;
        orientation = MapDirection.NORTH;
    }

    public Animal(){
        this(new Vector2d(2,2));
    }

    @Override
    public String toString() {
        return switch (this.orientation){
            case NORTH -> "^";
            case SOUTH -> "v";
            case WEST -> "<";
            case EAST -> ">";
        };
    }

    protected boolean isAt(Vector2d position){
        return this.position.equals(position);
    }

    public void move(MoveDirection direction, MoveValidator validator){
        Vector2d potentialMove = null;
        switch (direction){
            case MoveDirection.RIGHT -> setOrientation(orientation.next());
            case MoveDirection.LEFT -> setOrientation(orientation.previous());
            case MoveDirection.BACKWARD -> potentialMove = position.subtract(orientation.toUnitVector());
            case MoveDirection.FORWARD -> potentialMove = position.add(orientation.toUnitVector());
        }
        if(validator.canMoveTo(potentialMove))
            setPosition(potentialMove);
    }

    @Override
    public boolean equals(Object other){
        if(this==other) return true;
        if (other == null || getClass() != other.getClass())
            return false;
        Animal animal = (Animal) other;
        return (orientation.toUnitVector().equals(animal.orientation.toUnitVector()) && position.equals(((Animal) other).position));
    }

    @Override
    public int hashCode(){
        return Objects.hash(orientation, position);
    }
}

File: ProjektKotlina/src/main/kotlin/model/Animal.kt
package model

import java.util.*

class Animal(
    initialPosition: Vector2d,
    private var orientation: MapDirection
) : WorldElement {

    private var position: Vector2d = initialPosition
        private set

    fun getOrientation(): MapDirection = orientation

    fun setPosition(position: Vector2d) {
        this.position = position
    }

    override fun getPosition(): Vector2d = position

    override fun toString(): String = when (orientation) {
        MapDirection.NORTH -> "^"
        MapDirection.SOUTH -> "v"
        MapDirection.WEST -> "<"
        MapDirection.EAST -> ">"
    }

    fun isAt(position: Vector2d): Boolean = this.position == position

    fun move(direction: MoveDirection, validator: MoveValidator) {
        val potentialMove = when (direction) {
            MoveDirection.RIGHT -> {
                orientation = orientation.next()
                null
            }
            MoveDirection.LEFT -> {
                orientation = orientation.previous()
                null
            }
            MoveDirection.FORWARD -> position + orientation.toUnitVector()
            MoveDirection.BACKWARD -> position - orientation.toUnitVector()
        }

        if (potentialMove != null && validator.canMoveTo(potentialMove)) {
            position = potentialMove
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Animal) return false
        return orientation == other.orientation && position == other.position
    }

    override fun hashCode(): Int = Objects.hash(orientation, position)
}
